src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -33,9 +33,15 @@ internal static unsafe int ReadAtOffset(SafeFileHandle handle, Span<byte> buffer
                 // The Windows implementation uses ReadFile, which ignores the offset if the handle
                 // isn't seekable.  We do the same manually with PRead vs Read, in order to enable
                 // the function to be used by FileStream for all the same situations.
-                int result = handle.CanSeek ?
-                    Interop.Sys.PRead(handle, bufPtr, buffer.Length, fileOffset) :
-                    Interop.Sys.Read(handle, bufPtr, buffer.Length);
+                int result = Interop.Sys.PRead(handle, bufPtr, buffer.Length, fileOffset);
```

Comment:
The issue we're facing is that CanSeek returns true but PRead fails.  However, is there ever a situation where CanSeek would be false but PRead at arbitrary offsets would succeed?  Seems like we should continue using Read always if CanSeek is false; otherwise we're going to be paying for a PRead that's always going to fail.

Similar question/comment for the write path.

Reply:
+1

`CanSeek` was added to the handle pread/pwrite support. I think we should replace it by something like `MaybeCanRandomAccess` which optimistically defaults to `true`, and gets set to `false` in the error path.

Reply:
There is an almost-approved api proposal to expose SafeFileHandle.CanSeek https://github.com/dotnet/runtime/issues/58370 so I don't think we should remove it.
I added another field that caches "supports random access" and its initial value is set to CanSeek.

Reply:
> Seems like we should continue using Read always if CanSeek is false

Done, for read and for write as well.

Reply:
> There is an almost-approved api proposal to expose SafeFileHandle.CanSeek #58370 so I don't think we should remove it.

I've added a comment to that proposal: https://github.com/dotnet/runtime/issues/58370#issuecomment-933197378.

I think we should remove `CanSeek` in this PR or when implementing https://github.com/dotnet/runtime/issues/58381.
It makes sense to cache it as part of `FileStream`.